UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,696 @@
+          PciRootBridgeInfo->RootBridge[index].PMem.Limit           = PcdGet32 (PcdPciReservedPMemLimit);
+          PciRootBridgeInfo->RootBridge[index].UID                  = index;
+          PciRootBridgeInfo->RootBridge[index].HID                  = EISA_PNP_ID (0x0A03);
+
+          Data32 = (UINT32 *)(PropertyPtr->Data);
+          for (Base = 0; Base < 16; Base = Base + 7) {
+            DEBUG ((DEBUG_INFO, "  Base :%x \n", Base));
+            MemType = Fdt32ToCpu (*(Data32 + Base));
+            if (((MemType) & (SS_64BIT_MEMORY_SPACE)) == SS_64BIT_MEMORY_SPACE) {
+              DEBUG ((DEBUG_INFO, "  To program 64 mm \n"));
+              PciRootBridgeInfo->RootBridge[index].MemAbove4G.Base  = Fdt32ToCpu (*(Data32 + Base + 2)) + LShiftU64 (Fdt32ToCpu (*(Data32 + Base + 1)), 32);
+              PciRootBridgeInfo->RootBridge[index].MemAbove4G.Limit = PciRootBridgeInfo->RootBridge[index].MemAbove4G.Base + LShiftU64 (Fdt32ToCpu (*(Data32 + Base + 5)), 32) +  Fdt32ToCpu (*(Data32 + Base + 6)) -1;
+            } else if (((MemType) & (SS_32BIT_MEMORY_SPACE)) == SS_32BIT_MEMORY_SPACE) {
+              DEBUG ((DEBUG_INFO, "  To program 32 mem \n"));
+              PciRootBridgeInfo->RootBridge[index].Mem.Base  = Fdt32ToCpu (*(Data32 + Base + 2));
+              PciRootBridgeInfo->RootBridge[index].Mem.Limit = PciRootBridgeInfo->RootBridge[index].Mem.Base + Fdt32ToCpu (*(Data32 + Base + 6)) -1;
+            } else if (((MemType) & (SS_IO_SPACE)) == SS_IO_SPACE) {
+              DEBUG ((DEBUG_INFO, "  To program Io\n"));
+              PciRootBridgeInfo->RootBridge[index].Io.Base  = Fdt32ToCpu (*(Data32 + Base + 2));
+              PciRootBridgeInfo->RootBridge[index].Io.Limit = PciRootBridgeInfo->RootBridge[index].Io.Base + Fdt32ToCpu (*(Data32 +  Base + 6)) -1;
+            }
+          }
+
+          DEBUG ((DEBUG_INFO, "RootBridgeCount %x, index :%x\n", RootBridgeCount, index));
+
+          DEBUG ((DEBUG_INFO, "PciRootBridge->Mem.Base %x, \n", PciRootBridgeInfo->RootBridge[index].Mem.Base));
+          DEBUG ((DEBUG_INFO, "PciRootBridge->Mem.limit %x, \n", PciRootBridgeInfo->RootBridge[index].Mem.Limit));
+
+          DEBUG ((DEBUG_INFO, "PciRootBridge->MemAbove4G.Base %llx, \n", PciRootBridgeInfo->RootBridge[index].MemAbove4G.Base));
+          DEBUG ((DEBUG_INFO, "PciRootBridge->MemAbove4G.limit %llx, \n", PciRootBridgeInfo->RootBridge[index].MemAbove4G.Limit));
+
+          DEBUG ((DEBUG_INFO, "PciRootBridge->Io.Base %llx, \n", PciRootBridgeInfo->RootBridge[index].Io.Base));
+          DEBUG ((DEBUG_INFO, "PciRootBridge->Io.limit %llx, \n", PciRootBridgeInfo->RootBridge[index].Io.Limit));
+          if (index > 0) {
+            index--;
+          }
+        }
+
+        if (AsciiStrCmp (TempStr, "bus-range") == 0) {
+          UINT16  *Data16;
+
+          DEBUG ((DEBUG_INFO, "  Found bus-range Property TempLen (%08X)\n", TempLen));
+
+          if (index == 0xFF) {
+            index = RootBridgeCount - 1;

```

Comment:
index starting from 0xFF is weird, I thought index should start with 0 and if RootBridgeCount ==1, it will finished with index=0 (single iteration array)

Reply:
line 617 to make sure index only initialized once.
and the device node was discovered from the last to first.
if we have 2 root bridge , the root bridge 1 will be found first then root bridge 0.

Reply:
Doesn't sound right to me. First time index used in line 570 is used with index 0xff? Also why this logic have to be so complicated (initialize index within bus-range etc)? I have a working code which I shared earlier where index gets initialized to 0 and starts filling up info structure from that index onwards. It seems like there is some special case here into consideration that I fail to understand.

Reply:
pci-rb loop has been updated to resolve index issue